test/cpp/util/spiffe_test_credentials.cc

```diff
@@ -0,0 +1,144 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "test/cpp/util/spiffe_test_credentials.h"
+#include <thread>
+#include "test/core/end2end/data/ssl_test_data.h"
+
+namespace grpc {
+namespace testing {
+
+class TestSyncTlsCredentialReload
+    : public ::grpc_impl::experimental::TlsCredentialReloadInterface {
+  // Sync implementation.
+  int Schedule(
+      ::grpc_impl::experimental::TlsCredentialReloadArg* arg) override {
+    GPR_ASSERT(arg != nullptr);
+    struct ::grpc_impl::experimental::TlsKeyMaterialsConfig::PemKeyCertPair
+        pem_key_cert_pair = {test_server1_key, test_server1_cert};
+    std::vector<
+        struct ::grpc_impl::experimental::TlsKeyMaterialsConfig::PemKeyCertPair>
+        pem_key_cert_pair_list = {pem_key_cert_pair};
+    arg->set_key_materials(test_root_cert, pem_key_cert_pair_list);
+    arg->set_status(GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_NEW);
+    return 0;
+  }
+};
+
+class TestSyncTlsServerAuthorizationCheck
+    : public ::grpc_impl::experimental::TlsServerAuthorizationCheckInterface {
+  // Sync implementation.
+  int Schedule(
+      ::grpc_impl::experimental::TlsServerAuthorizationCheckArg* arg) override {
+    GPR_ASSERT(arg != nullptr);
+    arg->set_success(1);
+    arg->set_status(GRPC_STATUS_OK);
+    return 0;
+  }
+};
+
+static void TestAsyncTlsServerAuthorizationCheckCallback(
+    ::grpc_impl::experimental::TlsServerAuthorizationCheckArg* arg) {
+  GPR_ASSERT(arg != nullptr);
+  arg->set_success(1);
+  arg->set_status(GRPC_STATUS_OK);
+  arg->OnServerAuthorizationCheckDoneCallback();
+}
+
+class TestAsyncTlsServerAuthorizationCheck
+    : public ::grpc_impl::experimental::TlsServerAuthorizationCheckInterface {
+ public:
+  ~TestAsyncTlsServerAuthorizationCheck() {
+    if (thread_started_) {
+      server_authz_thread_.join();
+    }
+  }
+
+  // Async implementation.
+  int Schedule(
+      ::grpc_impl::experimental::TlsServerAuthorizationCheckArg* arg) override {
+    GPR_ASSERT(arg != nullptr);
+    thread_started_ = true;
+    server_authz_thread_ =
+        std::thread(TestAsyncTlsServerAuthorizationCheckCallback, arg);
+    return 1;
+  }
+
+  std::thread* server_authz_thread() { return &server_authz_thread_; }
+  bool* thread_started() { return &thread_started_; }
+
+ private:
+  std::thread server_authz_thread_;
+  bool thread_started_ = false;
+};
+
+TlsData CreateTestTlsCredentialsOptions(bool is_client, bool is_async) {
+  /** Create a credential reload config that is configured using the
+   *  TestSyncTlsCredentialReload class. **/
+  std::shared_ptr<TestSyncTlsCredentialReload> credential_reload_interface(
+      new TestSyncTlsCredentialReload());
+  std::shared_ptr<::grpc_impl::experimental::TlsCredentialReloadConfig>
+      test_credential_reload_config(
+          new ::grpc_impl::experimental::TlsCredentialReloadConfig(
+              credential_reload_interface));
+
```

Comment:
Please remove empty lines within the function.

Reply:
Done in next commit.